src/PostgREST/Response.hs

```diff
@@ -98,6 +96,9 @@ createResponse QualifiedIdentifier{..} MutateReadPlan{mrMutatePlan} ctxApiReques
     let
       pkCols = case mrMutatePlan of { Insert{insPkCols} -> insPkCols; _ -> mempty;}
       response = gucResponse rsGucStatus rsGucHeaders
+      prefHeader = prefAppliedHeader $
+        Preferences (if null pkCols && isNothing (qsOnConflict iQueryParams) then Nothing else preferResolution)
+        preferRepresentation Nothing preferCount preferTransaction preferMissing
```

Comment:
This logic would be better centralized on the `Plan` module, mostly ensuring that prefers make sense for the operation (e.g., `return=representation` doesn't make sense on GET). For now I'm leaving it here as it's already a good change.